### PR TITLE
[DependencyInjection] always call the parent class' constructor

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25343
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
This is a risky change: let's say the parent is customized: what if it has no constructor? or any args (less likely?)

**Comment by xabbuh** [comment]
Could happen, but right now we already call the parent constructor if the container is not compiled. Not doing this for compiled containers leads to inconsistencies and to issues as described in the linked issue.

**Comment by nicolas-grekas** [comment]
On the other hand, uncompiled container are not a thing, so their code path is mostly dead (we even removed it in 4.0.)

**Comment by xabbuh** [comment]
Hm indeed, you are right about that. Should we do some reflection then and try to find out if there is a constructor that could be called instead?

**Comment by nicolas-grekas** [comment]
Looks great yes. We should ensure that there is a constructor, and that it has zero mandatory args.

**Comment by nicolas-grekas** [comment]
On 3.4 only? I'd prefer not change any behavior on 2.*

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -970,9 +976,20 @@ public function __construct()
         }
 
         if ($this->container->isCompiled()) {
+            try {
+                $r = new \ReflectionClass($baseClassWithNamespace);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: $container->getReflectionClass()?

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 2]
```diff
@@ -970,9 +976,20 @@ public function __construct()
...
+
+                if ((null !== $constructor = $r->getConstructor()) && 0 === $constructor->getNumberOfRequiredParameters()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: && $baseClassWithNamespace !== Container::class

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: `if (($constructor = $r->getConstructor()) && !$constructor->getNumberOfRequiredParameters() && $baseClassWithNamespace !== Container::class) {`?

**Review by nicolas-grekas — APPROVED** [review]

**Review by 20uf — APPROVED** [review]

**Review by Taluu — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**